aiida_cp2k/calculations/datatype_helpers.py

```diff
@@ -0,0 +1,250 @@
+# -*- coding: utf-8 -*-
+###############################################################################
+# Copyright (c), The AiiDA-CP2K authors.                                      #
+# SPDX-License-Identifier: MIT                                                #
+# AiiDA-CP2K is hosted on GitHub at https://github.com/aiidateam/aiida-cp2k   #
+# For further information on the license, see the LICENSE.txt file.           #
+###############################################################################
+"""AiiDA-CP2K Gaussian Datatype Helpers."""
+
+import io
+from collections.abc import Sequence
+
+from aiida.common import InputValidationError
+from aiida.plugins import DataFactory
+
+
+def _unpack(adict):
+    """Unpack any lists as values into single elements for the key"""
+
+    for key, value in adict.items():
+        if isinstance(value, Sequence):
+            for item in value:
+                yield (key, item)
+        else:
+            yield (key, value)
+
+
+def _identifier(gdt):
+    """Our unique identifier for gaussian datatypes"""
+    return gdt.element, gdt.name
+
+
+def _validate_gdt_namespace(entries, gdt_cls, attr):
+    """Common namespace validator for both basissets and pseudos"""
+
+    identifiers = []
+
+    for kind, gdt_instance in _unpack(entries):
+        if not isinstance(gdt_instance, gdt_cls):
+            return "invalid {attr} for '{kind}' specified".format(attr=attr, kind=kind)
+
+        identifier = _identifier(gdt_instance)
+
+        if identifier in identifiers:
+            # note: this should be possible for basissets with different versions
+            #       but at this point we should require some format for the key to match it
+            return "{attr} for kind {gdt_instance.element} ({gdt_instance.name}) specified multiple times".format(
+                attr=attr, gdt_instance=gdt_instance)
+
+        identifiers += [identifier]
+
+    return None
+
+
+def _write_gdt(inp, entries, folder, key, fname):
+    """inject <key>=<fname> into all FORCE_EVAL/DFT sections and write the entries to a file"""
+
+    for secpath, section in inp.param_iter(sections=True):
+        if secpath[-1].upper() == "DFT":
+            section[key] = fname
+
+    with io.open(folder.get_abs_path(fname), mode="w", encoding="utf-8") as fhandle:
+        for _, entry in _unpack(entries):
+            entry.to_cp2k(fhandle)
+
+
+def validate_basissets_namespace(basissets, _):
+    """A input_namespace validator to ensure passed down basis sets have the correct type."""
+    return _validate_gdt_namespace(basissets, DataFactory("gaussian.basisset"), "basis set")
+
+
+def validate_basissets(inp, basissets, structure):
```

Comment:
Also, I am tempted to suggest to not use this check. It appears to be quite a difficult one to maintain. Let's maybe work a bit more on the `Cp2kInput` class to make it more robust and with a nice/powerful API. Once this is done, we introduce those checks back but we base them on the new `Cp2kInput` API. What do you think?

Reply:
> I see a lot of code duplication with the `validate_pseudos` function below. Can we merge them? It would make a lot of sence to me.

I actually tried it once (similar to the `_validate_gdt_namespace`), but I ended up with highly parametrized code which was barely readable. and not much less code. Which is why opted for some duplication in the end instead.

Reply:
> Also, I am tempted to suggest to not use this check. It appears to be quite a difficult one to maintain. Let's maybe work a bit more on the `Cp2kInput` class to make it more robust and with a nice/powerful API. Once this is done, we introduce those checks back but we base them on the new `Cp2kInput` API. What do you think?

I don't like it for several reasons:

* for the fully automated case where the user does not specify a `KIND` section anymore we have to generate the missing input at some place, hence we need some mechanism for that. Side-note: This could be done in `Cp2kInput` (or the respective `cp2k-input-tools`), but is also something which could be refactored later.
* for the complex manual assignment cases we should be validating to avoid bogus links in the provenance graph and ensure CP2K is picking up only the basissets/pseudos passed by AiiDA.
* loosening constraints (as done by these checks) should always be possible without breaking existing user-side code, while the other way round is most often not. So I would rather like to have it more complex on the developer side for now but safer and easier for the user to do the right thing, rather than the other way round

Reply:
>     * or the fully automated case where the user does not specify a `KIND` section anymore we have to generate the missing input at some place, hence we need some mechanism for that. Side-note: This could be done in `Cp2kInput` (or the respective `cp2k-input-tools`), but is also something which could be refactored later.

Will it work for multiple `FORCE_EVAL` sections?

Reply:
I definitely agree that such a check is important. I am just worried that it may limit the generality of the plugin. My secondary concern is that we may need to reimplement things when we move to the [cp2k-input-tools](https://github.com/cp2k/cp2k-input-tools). I didn't try it yet, but I believe you have a nice navigation mechanism there.

Reply:
> Will it work for multiple `FORCE_EVAL` sections? That was my primary concern. See an example [here](https://github.com/aiidateam/aiida-cp2k/blob/develop/examples/single_calculations/example_multiple_force_eval.py)

In the "manual" mode (user specifies basisset/pseudo names explicitly and passes the respective objects as inputs): yes. Here the validation is:
* for all KIND sections, see whether we have the basisset/pseudo in the basissets/pseudos inputs
* for all passed basissets/pseudos verify that they have been referenced in a KIND section

In the fully automated mode: not yet. The difficulty his that we might have to parse the fragments and the types of the force_evals to figure out where to create a subsys/kind section and for which symbols.

The problem is that without those checks, the user can essentially pass in whatever and it may or may not work depending on what is present on the target node, which gives a false sense of security and bogus provenance graphs at best, and calculations with unintended basissets/pseudos at worst.

Reply:
Tests for multiple FORCE_EVAL added.